build.gradle

```diff
@@ -95,31 +99,68 @@ task generateJniInterfaces(type:Exec) {
     environment 'CLASSPATH', sourceSets.main.runtimeClasspath.collect { it.path }.join(':')
     // JNI shared object isn't available at this stage
     environment 'WASMTIME_JNI_LOAD_DISABLED', '1'
-    commandLine(["${project.buildDir}/tmp/javah4x-${javah4xVersion}", 'rust', './wasmtime-jni/src'] + targets)
-    // Manually compare the source (class files) and output (rust sources) to run this task only when any
-    // update is made on classes.
-    outputs.upToDateWhen {
-        def smallestGeneratedTimestamp = targets.collect {
-            new File(project.rootDir, "wasmtime-jni/src/${it.replace(".", "_")}/mod.rs").lastModified()
-        }.min()
-        def largestSourceTimestamp = targets.collect {
-            def path = "classes/java/main/" + it.replace(".", "/") + ".class"
-            new File(project.buildDir, path).lastModified()
-        }.max()
-        smallestGeneratedTimestamp >= largestSourceTimestamp
+    if(jniLibOsClassifier().equals("windows")){
+        File file = new File("./generateJniInterfacesBat.bat")
+        file.createNewFile()
+        file.deleteOnExit()
+        String cmd = ""
+        for(String each:(["cmd","java -cp ${project.buildDir}\\tmp;" +
```

Comment:
Ah yes, we can use a downloaded executable as a JAR file in classpath, nice catch.
Then I think we can use `type:JavaExec` instead of branching how to run it on windows/unix, so gradle takes care of platform difference and  can run the main class of javah4x.
The following snippet worked perfectly on unix. Please try it on windows as well:
```groovy
task generateJniInterfaces(type:JavaExec) {
    def targets = [
            'io.github.kawamuray.wasmtime.Engine',
            'io.github.kawamuray.wasmtime.Func',
            'io.github.kawamuray.wasmtime.Instance',
            'io.github.kawamuray.wasmtime.Linker',
            'io.github.kawamuray.wasmtime.Memory',
            'io.github.kawamuray.wasmtime.Module',
            'io.github.kawamuray.wasmtime.Store',
            'io.github.kawamuray.wasmtime.wasi.Wasi']
    dependsOn 'unzipJavah4x', 'compileJava'
    // JNI shared object isn't available at this stage
    environment 'WASMTIME_JNI_LOAD_DISABLED', '1'
    // The downloaded executable is actually a shell script followed by JAR content.
    // Avoid executing it as a binary and use it as a JAR in classpath to make it work not only on Unix platform
    // but also on Windows.
    classpath "${project.buildDir}/tmp/javah4x-${javah4xVersion}"
    // Giving `sourceSets.main.runtimeClasspath` directly causes implicit dependency for `classes` task to be
    // added from this task, causing cyclic dependency. We only need compiled java classes so converting
    // existing entries to path and collect so it doesn't adds extra dependency.
    classpath sourceSets.main.runtimeClasspath.collect { it.path }
    main = 'javah4x.Javah4x'
    args(['rust', './wasmtime-jni/src'] + targets)
    // Manually compare the source (class files) and output (rust sources) to run this task only when any
    // update is made on classes.
    outputs.upToDateWhen {
        def smallestGeneratedTimestamp = targets.collect {
            new File(project.rootDir, "wasmtime-jni/src/${it.replace(".", "_")}/mod.rs").lastModified()
        }.min()
        def largestSourceTimestamp = targets.collect {
            def path = "classes/java/main/" + it.replace(".", "/") + ".class"
            new File(project.buildDir, path).lastModified()
        }.max()
        smallestGeneratedTimestamp >= largestSourceTimestamp
    }
}
```

Reply:
@kawamuray I find that this will change all files which named mod.rs from CRLF mode into LF mode and cause an exception, then the build failed. Is there a method to fix this problem?

Reply:
Because `javah4x` uses `\n` as the line separator, I think the generated `mod.rs` file always contain LF only I assume.
Do you mean it modifies the files that you manually changed to CRLF back into LF only?
Anyways, can you share the error you got while building the project?

Reply:
> Do you mean it modifies the files that you manually changed to CRLF back into LF only?  
Yes. But now there are no errors but if there are non-ascii characters, they won't display in the correct way.

Reply:
Hmm, how can a non-ascii character appear in the generated `mod.rs`?  I thought it consisting entirely with ASCII characters. Or maybe I'm misunderstanding your problem?
Anyway, I'd like to make build configuration common among different platforms as possible as we can, so I hope we can fix that problem and go with `JavaExec` which doesn't depend on any platform-specific premises.
Can you share the actual command you ran, and the output you got including the error messages?

btw, my build attempt with `JavaExec` at https://github.com/kawamuray/wasmtime-java/pull/3/files#diff-49a96e7eea8a94af862798a45174e6ac43eb4f8b4bd40759b5da63ba31ec3ef7R88 successfully passed the build on windows, so I believe it'll work well on windows environment too.

Reply:
It can successfully build. But the terminal prints like this:
![image](https://user-images.githubusercontent.com/51087818/102344488-eefc5e80-3fd6-11eb-9b75-78b3b2f8396c.png)

Reply:
I see... interesting. However I'm very sure that it is not related to the generated files, since I could confirm those files are totally consisting with ASCII chars:

```sh
$ nkf --guess wasmtime-jni/src/*/mod.rs
wasmtime-jni/src/io_github_kawamuray_wasmtime_Engine/mod.rs: ASCII (LF)
wasmtime-jni/src/io_github_kawamuray_wasmtime_Func/mod.rs: ASCII (LF)
wasmtime-jni/src/io_github_kawamuray_wasmtime_Instance/mod.rs: ASCII (LF)
wasmtime-jni/src/io_github_kawamuray_wasmtime_Linker/mod.rs: ASCII (LF)
wasmtime-jni/src/io_github_kawamuray_wasmtime_Memory/mod.rs: ASCII (LF)
wasmtime-jni/src/io_github_kawamuray_wasmtime_Module/mod.rs: ASCII (LF)
wasmtime-jni/src/io_github_kawamuray_wasmtime_Store/mod.rs: ASCII (LF)
wasmtime-jni/src/io_github_kawamuray_wasmtime_wasi_Wasi/mod.rs: ASCII (LF)
```

I've also checked if any files in the project contains any multibyte characters but apparently not except those binaries misdetected:
```sh
$ find . -type f | grep -v '\.git' | grep -v './build' | grep -v './wasmtime-jni/target' | xargs nkf --guess | grep -v -E 'ASCII|BINARY'
./.gradle/6.7.1/javaCompile/javaCompile.lock: EUC-JP
./.gradle/6.7.1/fileHashes/fileHashes.lock: CP932 (CR)
./.gradle/6.6.1/fileContent/fileContent.lock: CP932
./.gradle/6.6.1/javaCompile/javaCompile.lock: Shift_JIS
./.gradle/6.6.1/fileHashes/fileHashes.lock: Shift_JIS
```

So it seems a specific issue of Intellij on windows. Actually a bit of ggling taught me that it is a common problem for intellij/windows users. [This post](https://qiita.com/SierSetup/items/14314c8b5c0a932cfda2#gradle%E3%81%AE%E3%82%B3%E3%83%B3%E3%83%91%E3%82%A4%E3%83%AB%E6%99%82%E3%81%AE%E3%82%A8%E3%83%B3%E3%82%B3%E3%83%BC%E3%83%89) (Japanese) suggested two solutions, one is [this](https://www.jetbrains.com/help/idea/configuring-individual-file-encoding.html#console) and the another is adding below in build.gradle. Does that possibly helps you?
```
tasks.withType(JavaCompile) {
    options.encoding = 'UTF-8'
}
```

Reply:
I think this is a bug of Intellij IDEA. It doesn't matter when I change the build terminal from the one inside Intellij IDEA to the system terminal, the problem is solved.